### ISSUE TITLE
Got AssertionError with Doc2Vec

### ISSUE THREAD

**Issue by tinayliu** [RaRe-Technologies/gensim#715]
Hi,

I got AssertionError when running the following code snippet to train a doc2vec model. 

from gensim.models.doc2vec import Doc2Vec, TaggedDocument

```
class TaggedDocumentsIter(object):
    """Iterate over each line in the file, yielding each document out as a TaggedDocument object."""
    def __init__(self, filename):
        self.filename = filename

    def __iter__(self):
        for item_no, line in enumerate(utils.smart_open(self.filename)):
            tokens = line.strip().split(' ')
            label = tokens[0]
            tokens.remove(label)
            yield TaggedDocument(words=tokens, tags=[label])

documents = TaggedDocumentsIter( data_file )
model = Doc2Vec(documents, size=300, min_count=1, workers=6, window=5,  sample=0.001, alpha=0.025, negative=5)
```

The outputs from above code:

```
2016-06-01 19:11:48,475 : MainThread : INFO : collecting all words and their counts
2016-06-01 19:11:48,580 : MainThread : INFO : PROGRESS: at example #0, processed 0 words (0/s), 0 word types, 0 tags
2016-06-01 19:11:57,593 : MainThread : INFO : PROGRESS: at example #10000, processed 23177033 words (2571738/s), 1178622 word types, 10000 tags
2016-06-01 19:12:06,658 : MainThread : INFO : PROGRESS: at example #20000, processed 46988814 words (2627003/s), 1961980 word types, 20000 tags
2016-06-01 19:12:13,009 : MainThread : INFO : PROGRESS: at example #30000, processed 63768819 words (2642701/s), 2408922 word types, 28381 tags
2016-06-01 19:12:19,785 : MainThread : INFO : PROGRESS: at example #40000, processed 80055679 words (2403618/s), 2855149 word types, 38087 tags
...

2016-06-01 19:37:45,830 : MainThread : INFO : PROGRESS: at example #15890000, processed 3129855878 words (1918890/s), 49205045 word types, 15816657 tags
2016-06-01 19:37:46,313 : MainThread : INFO : PROGRESS: at example #15900000, processed 3130776370 words (1909040/s), 49218220 word types, 15826589 tags
2016-06-01 19:37:46,373 : MainThread : INFO : collected 49219779 word types and 15827710 unique tags from a corpus of 15901127 examples and 3130888965 words
2016-06-01 19:45:25,614 : MainThread : INFO : min_count=1 retains 49219779 unique words (drops 0)
2016-06-01 19:45:25,615 : MainThread : INFO : min_count leaves 3130888965 word corpus (100% of original 3130888965)
2016-06-01 19:49:50,726 : MainThread : INFO : deleting the raw counts dictionary of 49219779 items
2016-06-01 19:49:54,529 : MainThread : INFO : sample=0.0001 downsamples 353 most-common words
2016-06-01 19:49:54,530 : MainThread : INFO : downsampling leaves estimated 2132412884 word corpus (68.1% of prior 3130888965)
2016-06-01 19:49:54,530 : MainThread : INFO : estimated required memory for 49219779 words and 300 dimensions: 233803843700 bytes
2016-06-01 19:51:48,546 : MainThread : INFO : constructing a huffman tree from 49219779 words
2016-06-01 20:53:41,769 : MainThread : INFO : built huffman tree with maximum node depth 32
Traceback (most recent call last):
  File "doc2vec.py", line 128, in <module>
    # http://radimrehurek.com/gensim/models/doc2vec.html#blog
  File "/usr/local/lib/python2.7/site-packages/gensim/models/doc2vec.py", line 596, in __init__
    self.build_vocab(documents, trim_rule=trim_rule)
  File "/usr/local/lib/python2.7/site-packages/gensim/models/word2vec.py", line 497, in build_vocab
    self.finalize_vocab()  # build tables & arrays
  File "/usr/local/lib/python2.7/site-packages/gensim/models/word2vec.py", line 618, in finalize_vocab
    self.make_cum_table()
  File "/usr/local/lib/python2.7/site-packages/gensim/models/word2vec.py", line 455, in make_cum_table
    assert self.cum_table[-1] == domain
AssertionError
```

Any help is greatly appreciated.

Thanks,
Tina

**Comment by gojomo** [comment]
Probably caused by 32bit Python: 3+ trillion word corpus whose sampling-splits can't be stored in 32bit int cum_table array. See my forum post: https://groups.google.com/d/msg/gensim/AHRrlnZPYbg/1LmXkTvPOwAJ

We could potentially give a better error message - but in general working with this size corpus (or any large word2vec/doc2vec models) will require 64bit Python.

**Comment by tinayliu** [comment]
Hi Gordon,

Thank you for the quick reply.

I was already using 64-bits python. 
$ python -c 'import struct;print( 8 * struct.calcsize("P"))'
64

I got the same error with a much smaller dataset.

2016-06-02 22:01:50,244 : MainThread : INFO : collecting all words and their counts
2016-06-02 22:01:50,245 : MainThread : INFO : PROGRESS: at example #0, processed 0 words (0/s), 0 word types, 0 tags
2016-06-02 22:01:58,211 : MainThread : INFO : PROGRESS: at example #10000, processed 23228924 words (2916156/s), 1180387 word types, 10000 tags
2016-06-02 22:02:06,416 : MainThread : INFO : PROGRESS: at example #20000, processed 47020438 words (2900200/s), 1962690 word types, 20000 tags
2016-06-02 22:02:13,090 : MainThread : INFO : PROGRESS: at example #30000, processed 66315589 words (2891350/s), 2483088 word types, 29930 tags
2016-06-02 22:02:19,108 : MainThread : INFO : PROGRESS: at example #40000, processed 82294873 words (2655715/s), 2922443 word types, 39928 tags
...
2016-06-02 22:05:26,029 : MainThread : INFO : PROGRESS: at example #860000, processed 553075769 words (2513022/s), 11919966 word types, 859680 tags
2016-06-02 22:05:27,692 : MainThread : INFO : PROGRESS: at example #870000, processed 557259919 words (2516113/s), 11976280 word types, 869676 tags
2016-06-02 22:05:27,861 : MainThread : INFO : collected 11982321 word types and 870629 unique tags from a corpus of 870953 examples and 557688958 words
2016-06-02 22:06:52,117 : MainThread : INFO : min_count=1 retains 11982321 unique words (drops 0)
2016-06-02 22:06:52,118 : MainThread : INFO : min_count leaves 557688958 word corpus (100% of original 557688958)
2016-06-02 22:07:50,340 : MainThread : INFO : deleting the raw counts dictionary of 11982321 items
2016-06-02 22:07:50,789 : MainThread : INFO : sample=0.001 downsamples 25 most-common words
2016-06-02 22:07:50,789 : MainThread : INFO : downsampling leaves estimated 463373524 word corpus (83.1% of prior 557688958)
2016-06-02 22:07:50,790 : MainThread : INFO : estimated required memory for 11982321 words and 300 dimensions: 52742860900 bytes
2016-06-02 22:08:11,708 : MainThread : INFO : constructing a huffman tree from 11982321 words
2016-06-02 22:21:51,869 : MainThread : INFO : built huffman tree with maximum node depth 29
Traceback (most recent call last):
  File "doc2vec_standalone.py", line 131, in <module>
    sample=args.sample, alpha=args.alpha, negative=args.negative)
  File "/usr/local/lib/python2.7/site-packages/gensim/models/doc2vec.py", line 596, in __init__
    self.build_vocab(documents, trim_rule=trim_rule)
  File "/usr/local/lib/python2.7/site-packages/gensim/models/word2vec.py", line 497, in build_vocab
    self.finalize_vocab()  # build tables & arrays
  File "/usr/local/lib/python2.7/site-packages/gensim/models/word2vec.py", line 618, in finalize_vocab
    self.make_cum_table()
  File "/usr/local/lib/python2.7/site-packages/gensim/models/word2vec.py", line 455, in make_cum_table
    assert self.cum_table[-1] == domain
AssertionError

I did successfully trained with a very similar dataset. I am using gensim-0.12.3. Is there any related bug fix since then? 

Thanks,
Tina

**Comment by gojomo** [comment]
Hmm. 

I don't recall a specific fix, but I would definitely try the latest released version. Note, though, that in the latest version you may need to explicitly state `hs=1` to cause simultaneous HS & negative-sampling training, to match what you're currently doing. (It's not typical to use both simultaneously, but it should still work.)

Could you print the value of `cum_table[-1]` after the error is reached? 

(Looking at the sampling-table, I am now wondering if there are cases with a large number of words where it breaks or some words can't possibly be sampled... at least not without a further adjustment to the randomization.)

**Comment by tinayliu** [comment]
Hi Gordon,

I got the same error with the latest gensim. The value of cum_table[-1] is '2147483648' right before the assertion.

Thanks,
Tina

**Comment by gojomo** [comment]
Aha. So the value which is expected to be (2**31 - 1) is instead 2**31. I suspect there's a bit of floating-point imprecision which only comes into play for certain corpus sizes – maybe extra-large ones, maybe those of particular distributions, maybe (less likely) on just some OS/python/CPU combos.

Not yet sure of the right fix. 

@tinayliu - for my curiosity, what are the 'word types', 'unique tags', and corpus 'examples' and 'words', counts for your data from INFO logging? How about the "word corpus" size after applying min_count and downsampling?

**Comment by tinayliu** [comment]
Gordon,

The last few lines in the logging are:

collected 11982321 word types and 870629 unique tags from a corpus of 870953 examples and 557688958 words
min_count=1 retains 11982321 unique words (drops 0)
min_count leaves 557688958 word corpus (100% of original 557688958)
deleting the raw counts dictionary of 11982321 items
sample=0.001 downsamples 25 most-common words
downsampling leaves estimated 463373524 word corpus (83.1% of prior 557688958)
estimated required memory for 11982321 words and 300 dimensions: 52742860900 bytes
constructing a huffman tree from 11982321 words
built huffman tree with maximum node depth 29

Thanks,
Tina

**Comment by tinayliu** [comment]
Gordon,

I can provide the data that caused the assertion for you to debug. The data file is 1.2G after compressed. if you are interested in it, please let me know how I send it to you.

Thanks,
Tina

**Comment by tmylk** [comment]
Duplicate of https://github.com/RaRe-Technologies/gensim/issues/865 resolved in #876
